### ISSUE TITLE
Cubemap orientation

### ISSUE THREAD

**Issue by fabienrohrer** [cyberbotics/webots#387]
Following #375, the background orientation (and so environment map reflections) in THREEjs and Webots are rotated by 180° vertically. This is currently solved in JS by flipping the top/bottom images and inverting the right/left and front/back textures.

We could explore now a better solution: to break the Webots background orientation.
First step is to determine our compliance with X3D to determine if this is a good idea or not.

**Comment by fabienrohrer** [comment]
The Webots orientation does respect the VRML and X3D Background definition, which is good:

As a proof ("right" == `+x`, "back" == `+z`):

![back](https://user-images.githubusercontent.com/866788/56721431-286ac100-6745-11e9-996c-e51a4b01a055.png)
![right](https://user-images.githubusercontent.com/866788/56721432-286ac100-6745-11e9-9347-cd56de8b8b9d.png)

Therefore, it's a good idea to keep the Webots background orientation as-is, otherwise we will have future issues.

So to improve this topic, it remains only the possibility to modify THREEjs.
I will start to ask a question on to the THREEjs Github issues.

**Comment by fabienrohrer** [comment]
Follow up: https://github.com/mrdoob/three.js/issues/16328

**Comment by fabienrohrer** [comment]
I tried to put hands on the `three.js` sources.
I can change the background and reflection cubemap orientation with this patch:

```diff
diff --git a/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js b/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js
index d319e9b88..27dd325a1 100644
--- a/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js
+++ b/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js
@@ -72,7 +72,8 @@ export default /* glsl */`
 
                #ifdef ENVMAP_TYPE_CUBE
 
-                       vec3 queryReflectVec = vec3( flipEnvMap * reflectVec.x, reflectVec.yz );
+                       mat3 m = mat3(1.0, 0.0, 0.0, 0.0, 0.0, -1.0, 0.0, 1.0, 0.0);
+                       vec3 queryReflectVec = m * vec3( flipEnvMap * reflectVec.x, reflectVec.yz );

diff --git a/src/renderers/shaders/ShaderLib/cube_frag.glsl.js b/src/renderers/shaders/ShaderLib/cube_frag.glsl.js
index 0a53ba08e..dc7ba4699 100644
--- a/src/renderers/shaders/ShaderLib/cube_frag.glsl.js
+++ b/src/renderers/shaders/ShaderLib/cube_frag.glsl.js
@@ -6,8 +6,8 @@ uniform float opacity;
 varying vec3 vWorldDirection;
 
 void main() {
-
-       vec4 texColor = textureCube( tCube, vec3( tFlip * vWorldDirection.x, vWorldDirection.yz ) );
+       mat3 m = mat3(1.0, 0.0, 0.0, 0.0, 0.0, -1.0, 0.0, 1.0, 0.0);
+       vec4 texColor = textureCube( tCube, m * vec3( tFlip * vWorldDirection.x, vWorldDirection.yz ) );
```

<img width="690" alt="Capture d’écran 2019-05-21 à 10 49 27" src="https://user-images.githubusercontent.com/866788/58083150-bb6a1000-7bb8-11e9-9fbd-d1fc659ac79e.png">

**Comment by stefaniapedrazzi** [comment]
I just checked and there is no Cubemap orientation issues when using an equirectangular image (default backgrounds in Webots).

**Comment by fabienrohrer** [comment]
Great.

**Comment by stefaniapedrazzi** [comment]
Or maybe not, I'm still checking..

**Comment by fabienrohrer** [comment]
This information is important. I could also fix it if I do a threejs patch.

**Comment by stefaniapedrazzi** [comment]
I'm still having issues applying the background so it doesn't work in webots.min.js, but only in a separate three.js example that doesn't work the same for the moment.
I will confirm if it is correct when it will fully work.

**Comment by stefaniapedrazzi** [comment]
Also for HDR equirectangular background, the textures we distribute are mirrored horizontally when applied to the three.js scene.
HDR equirectangular textures from the three.js repository are rendered correctly.
In this case we have to change the shaders defined in the `EquirectangularToCubeGenerator.js` example source file and it would also be easier to patch it locally given that it has to be minified and bundled within webots.min.js.

Note that HDR images always need to be flipped vertically to be displayed correctly.

**Comment by BenjaminDeleze** [comment]
Fixed by using wrenjs instead of threejs #2769
